### DISCUSSION TITLE
vim.lsp.start on FileType autocmd does not need root_dir?

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [neovim/neovim#31767] [Q&A] [answered]
I'm experimenting with replacing lspconfig with the new native `vim.lsp` functions.

I have this in a `FileType` autocmd:

```lua
--- @param server string The LSP name.
--- @param server_opts vim.lsp.Config The LSP config.
local function start_lsp(server, server_opts)
	vim.lsp.config[server] = server_opts
	
	---@type vim.lsp.ClientConfig
	local client_config = { name = server, cmd = cmd }
	
	---@type vim.lsp.start.Opts
	local start_opts = { bufnr = 0, attach = true }

	vim.lsp.start(client_config, start_opts)
end
```

The autocmd can then attach the buf to the started LSP. This seems to work great as long as `root_markers` are part of the `server_opts` spec.

My question: when reading the docs, it seems like I am expected to pass in `root_dir` into the `vim.lsp.ClientConfig` above; `{ name = server, cmd = cmd, root_dir = root_dir }` but this causes issues with e.g. the basedpyright LSP which refuses to start if I do so.

Am I doing this right, that I am relying on the `root_markers` rather than the `root_dir`?

Full context in this PR: https://github.com/fredrikaverpil/dotfiles/pull/167/files

**Comment by clason** [comment]
Don't mix `vim.lsp.start` with `vim.lsp.config` -- the latter is a higher-level API that replaces `vim.lsp.start`.

**Reply by fredrikaverpil** [reply]
Aha, I see. If only using `vim.lsp.config`, how would I start the LSP?
Is it `vim.lsp.enable`?

```lua
vim.lsp.config[server] = server_opts
vim.lsp.enable(server)
```

Does this also mean I won't be needing an autocmd anymore?

**Reply by clason** [reply, marked as answer]
Yes, and yes.

**Reply by fredrikaverpil** [reply]
Awesome, many thanks!
